crates/sound/src/audio_backends/pw_backend.rs

```diff
@@ -160,13 +198,189 @@ impl AudioBackend for PwBackend {
+                        let req = (*b).requested * (frame_size as u64);
+                        let mut n_bytes = cmp::min(req as u32, (*datas).maxsize);
+
+                        let buffer: Vec<u8> = vec![0; RINGBUFFER_SIZE as usize];
+                        let buffer_ptr = buffer.as_ptr() as *const c_void;
+
+                        //use spa_ringbuffer functions
+                        let mut index: u32 = 0;
+                        let mut ring = spa_ringbuffer {
+                            readindex: 0,
+                            writeindex: 0,
+                        };
+                        //get no of available bytes to read data from buffer
+                        let avail = spa_ringbuffer_get_read_index(&mut ring, &mut index);
+                        if avail < n_bytes as i32 {
+                            n_bytes = avail.try_into().unwrap();
+                        }
+                        spa_ringbuffer_read_data(
+                            &mut ring,
+                            buffer_ptr,
+                            RINGBUFFER_SIZE,
+                            index & RINGBUFFER_MASK,
+                            p,
+                            n_bytes,
+                        );
+                        index += n_bytes;
+                        spa_ringbuffer_read_update(&mut ring, index);
+
+                        (*(*datas).chunk).offset = 0;
+                        (*(*datas).chunk).stride = frame_size as i32;
+                        (*(*datas).chunk).size = n_bytes;
+                        stream.queue_raw_buffer(b);
+                    }
+                })
+                .register()
+                .expect("failed to register stream listener");
+
+            stream_listener.insert(stream_id, listener_stream);
+
+            let direction = match params.direction {
+                VIRTIO_SND_D_OUTPUT => spa::Direction::Output,
+                VIRTIO_SND_D_INPUT => spa::Direction::Input,
+                _ => panic!("Invalid direction"),
+            };
+
+            stream
+                .connect(
+                    direction,
+                    Some(pipewire::constants::ID_ANY),
+                    pw::stream::StreamFlags::RT_PROCESS
+                        | pw::stream::StreamFlags::AUTOCONNECT
+                        | pw::stream::StreamFlags::MAP_BUFFERS,
+                    &mut [param],
+                )
+                .expect("could not connect to the stream");
+
+            self.thread_loop.unlock();
+            // insert created stream in a hash table
+            stream_hash.insert(stream_id, stream);
+
+            // state {SET_PARAMS} -> {PREPARE}
+            stream_states[stream_id as usize] = VIRTIO_SND_R_PCM_PREPARE;
+        } else if stream_states[stream_id as usize] == VIRTIO_SND_R_PCM_PREPARE {
+            let stream = stream_hash.get(&stream_id).unwrap();
+            stream
+                .update_params(&mut [param])
```

Comment:
I think here you have to rebuilt `info` to get `param` again like:
```
let param: *mut spa_pod =
            unsafe { spa_format_audio_raw_build(&mut b, SPA_PARAM_EnumFormat, &mut info) };
```
Otherwise, it will contain the old values of the stream.